clang/lib/Sema/SemaChecking.cpp

```diff
@@ -11626,10 +11626,14 @@ void Sema::CheckImplicitConversion(Expr *E, QualType T, SourceLocation CC,
                   Context.getFloatTypeSemantics(QualType(TargetBT, 0)),
                   Context.getFloatTypeSemantics(QualType(SourceBT, 0)))) {
             if (getLangOpts().SYCLIsDevice)
-              SYCL().DiagIfDeviceCode(CC, diag::warn_imp_float_size_conversion);
+              SYCL().DiagIfDeviceCode(
+                  CC, diag::warn_sycl_imp_float_size_conversion);
             else
               DiagnoseImpCast(*this, E, T, CC,
-                              diag::warn_imp_float_size_conversion);
+                              getLangOpts().isSYCL()
```

Comment:
For non-SYCL mode, I think we can make the behavior similar to community clang here. So maybe just handle the SYCL case.  That way we can stick with just the one diagnostic of `warn_imp_float_size_conversion` i.e. 
```
if (getLangOpts().isSYCL())
     if (getLangOpts().SYCLIsDevice)
        SYCL().DiagIfDeviceCode(
                  CC, diag::warn_sycl_imp_float_size_conversion);
     else 
         DiagnoseImpCast(*this, E, T, CC,
             diag::warn_imp_float_size_conversion);
```